src/engine/Engine.h

```diff
@@ -367,8 +369,301 @@ class Engine {
     }
   }
 
-private:
+  template<typename R, int K>
+  struct newOptionalResult {
+    R operator()(unsigned int resultSize) {
+      (void) resultSize;
+      return R();
+    }
+  };
+
+  template<int K>
+  struct newOptionalResult<std::vector<Id>, K> {
+    std::vector<Id> operator()(unsigned int resultSize) {
+      return vector<Id>(resultSize);
+    }
+  };
+
+
+ template<typename A, typename B, typename R, bool aEmpty, bool bEmpty>
+  static void createOptionalResult(const typename A::value_type* a,
+                                   const typename B::value_type* b,
+                                   size_t sizeA,
+                                   int jcls_a, int jcls_b,
+                                   const std::vector<size_t>& jclAToB,
+                                   unsigned int resultSize,
+                                   R& res) {
+    assert(!(aEmpty && bEmpty));
+    if (aEmpty) {
+      // Fill the columns of a with ID_NO_VALUE and the rest with b.
+      size_t i = 0;
+      for (size_t col = 0; col < sizeA; col++) {
+        if ((jcls_a & (1 << col)) == 0) {
+          res[col] = ID_NO_VALUE;
+        } else {
+          // if this is one of the join columns use the value in b
+          res[col] = (*b)[jclAToB[col]];
+        }
+        i++;
+      }
+      for (size_t col = 0; col < b->size(); col++) {
+        if ((jcls_b & (1 << col)) == 0) {
+          // only write the value if it is not one of the join columns in b
+          res[i] = (*b)[col];
+          i++;
+        }
+      }
+    } else if (bEmpty) {
+      // Fill the columns of b with ID_NO_VALUE and the rest with a
+      for (size_t col = 0; col < sizeA; col++) {
+        res[col] = (*a)[col];
+      }
+      for (size_t col = sizeA; col < resultSize; col++) {
+        res[col] = ID_NO_VALUE;
+      }
+    } else {
+      // Use the values from both a and b
+      unsigned int i = 0;
+      for (size_t col = 0; col < a->size(); col++) {
+        res[col] = (*a)[col];
+        i++;
+      }
+      for (size_t col = 0; col < b->size(); col++) {
+        if ((jcls_b & (1 << col)) == 0) {
+          res[i] = (*b)[col];
+          i++;
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Joins two result tables on any number of columns, inserting the
+   *        special value ID_NO_VALUE for any entries marked as optional.
+   * @param a
+   * @param b
+   * @param aOptional
+   * @param bOptional
+   * @param jcls
+   * @param result
+   */
+  template<typename A, typename B, typename R, int K>
+  static void optionalJoin(const A& a, const B& b,
+                           bool aOptional, bool bOptional,
+                           const vector<array<size_t, 2>>& jcls,
```

Comment:
can you provide more desriptive names here? I have no idea what `jcls` stands for

Reply:
I replaced all mentions of jcl and jcls with joinColumn and joinColumns respectively.